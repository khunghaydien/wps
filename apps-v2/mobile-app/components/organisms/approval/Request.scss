@import 'mobile-global-imports.scss';

.mobile-app-organisms-approval-request {
  &__title {
    margin-bottom: $small-size;
    color: $color-blue-800;
    font-size: unit(5);
    line-height: 1;
  }
}
