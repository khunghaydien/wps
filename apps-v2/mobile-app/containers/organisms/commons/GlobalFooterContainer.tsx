import React, { useCallback, useMemo } from 'react';
import { RouteComponentProps } from 'react-router-dom';

import { $Values } from 'utility-types';

import { EXPENSE_APPROVAL_REQUEST } from '@mobile/constants/approvalRequest';

import msg from '@commons/languages';
import GlobalFooter from '@mobile/components/organisms/commons/GlobalFooter';

export const EXPENSE_TABS = {
  report: 'report',
  record: 'record',
} as const;

export const PAGE = {
  expenseApproval: 'expenseApproval',
};

const PATH = {
  approvalExpense: '/approval/expenses/list/',
  approvalRequest: '/approval/requests/list/',
};

type Props = {
  page: $Values<typeof PAGE>;
  history: RouteComponentProps['history'];
  activeTab: $Values<typeof EXPENSE_TABS & typeof EXPENSE_APPROVAL_REQUEST>;
  handleTabChange?: (string) => void;
};

const GlobalFooterContainer = (ownProps: Props) => {
  const onClickTab = useCallback(
    (url: string, tab?: string) => () => {
      if (ownProps.handleTabChange) {
        ownProps.handleTabChange(tab);
      }
      ownProps.history.replace(url);
    },
    [ownProps.history]
  );

  const tabs = useMemo(() => {
    let tabs;
    switch (ownProps.page) {
      case PAGE.expenseApproval:
        tabs = [
          {
            label: msg().Appr_Clbl_ExpensesRequest,
            handleOnClick: onClickTab(
              PATH.approvalExpense,
              EXPENSE_APPROVAL_REQUEST.expense
            ),
            key: EXPENSE_APPROVAL_REQUEST.expense,
          },
          {
            label: msg().Appr_Clbl_ExpensesPreApproval,
            handleOnClick: onClickTab(
              PATH.approvalRequest,
              EXPENSE_APPROVAL_REQUEST.request
            ),
            key: EXPENSE_APPROVAL_REQUEST.request,
          },
        ];
        break;
      default:
        tabs = [];
    }
    return tabs;
  }, [module]);

  return <GlobalFooter tabs={tabs} activeTab={ownProps.activeTab} />;
};

export default GlobalFooterContainer;
