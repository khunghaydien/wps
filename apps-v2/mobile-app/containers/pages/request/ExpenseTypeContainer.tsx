import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { RouteComponentProps } from 'react-router-dom';
import { AnyAction } from 'redux';
import { ThunkDispatch } from 'redux-thunk';

import _ from 'lodash';

import {
  goBack,
  pushHistoryWithPrePage,
} from '@mobile/concerns/routingHistory';

import {
  isMileageRecord,
  newRecord,
  RECEIPT_TYPE,
  RECORD_TYPE,
  RECORD_TYPE_CATEGORY,
} from '../../../../domain/models/exp/Record';
import { ExpenseType } from '@apps/domain/models/exp/ExpenseType';
import { isUseJctNo, JCT_NUMBER_INVOICE } from '@apps/domain/models/exp/JCTNo';
import { initialDestinations } from '@apps/domain/models/exp/Mileage';

import { State } from '../../../modules';
import { actions as routeFormPageActions } from '../../../modules/expense/pages/routeFormPage';
import { actions as formValueAction } from '../../../modules/expense/ui/general/formValues';
import { actions as mileageActions } from '@mobile/modules/expense/ui/mileage';

import {
  getExpenseTypeList,
  searchExpenseType,
} from '../../../action-dispatchers/expense/ExpenseType';
import { getPaymentMethodOptionList } from '@mobile/action-dispatchers/expense/PaymentMethod';

import ExpTypeComponent from '../../../components/pages/expense/Record/New/General/ExpenseType';

type OwnProps = RouteComponentProps & {
  type: string;
  recordType?: string;
  parentGroupId?: string;
  parentParentGroupId?: string;
  keyword?: string;
  level?: number;
};

const ExpenseTypeContainer = (ownProps: OwnProps) => {
  const dispatch = useDispatch() as ThunkDispatch<unknown, void, AnyAction>;

  const companyId = useSelector((state: State) => state.userSetting.companyId);
  const routeFormPage = useSelector(
    (state: State) => state.expense.pages.routeFormPage
  );
  const generalFormValues = useSelector(
    (state: State) => state.expense.ui.general.formValues
  );
  const expenseTypeList = useSelector(
    (state: State) => state.expense.entities.expenseTypeList.records
  );
  const hasMore = useSelector(
    (state: State) => state.expense.entities.expenseTypeList.hasMore
  );
  const accountingDate = useSelector((state: State) => {
    return state.expense.entities.report.scheduledDate;
  });
  const reportId = useSelector(
    (state: State) => state.expense.entities.report.reportId
  );
  const reportTypeId = useSelector(
    (state: State) => state.expense.entities.report.expReportTypeId
  );
  const employeeDetails = useSelector(
    (state: State) => state.common.exp.entities.employeeDetails
  );
  let subroleId = _.get(employeeDetails, 'selectedRole');
  subroleId = subroleId === 'primary' ? undefined : subroleId;

  const useJctRegistrationNumber = useSelector(
    (state: State) => state.userSetting.jctInvoiceManagement
  );

  const recordType =
    ownProps.recordType ||
    _.get(ownProps.history, 'location.state.recordType', '');

  const formValues =
    recordType === RECORD_TYPE.TransitJorudanJP
      ? routeFormPage
      : generalFormValues;

  useEffect(() => {
    const finalReportTypeId = reportTypeId;
    const formDate = accountingDate;

    const targetDate = formDate;
    if (ownProps.type === 'list') {
      const parentGroupId = ownProps.parentGroupId || null;
      dispatch(
        getExpenseTypeList(
          null,
          parentGroupId,
          targetDate,
          recordType,
          finalReportTypeId,
          true,
          subroleId
        )
      );
    } else {
      const keyword = ownProps.keyword || '';
      dispatch(
        searchExpenseType(
          companyId,
          decodeURIComponent(keyword),
          targetDate,
          recordType,
          finalReportTypeId,
          true
        )
      );
    }
  }, []);

  const onSearchClick = (keyword: string) => {
    if (keyword === null) {
      return;
    }
    let level = 2;
    if (ownProps.parentParentGroupId) {
      level = 4;
    } else if (ownProps.parentGroupId) {
      level = 3;
    }

    // Special handling of % because it could not be sanitized with react router.
    const sanitizedKeyword = encodeURIComponent(keyword.replace(/%/g, '%25'));
    const urlPath = `/request/expense-type/search/keyword=${sanitizedKeyword}/${level}`;
    const state = {
      target: _.get(ownProps.history, 'location.state.target'),
      recordType,
      recordId: _.get(ownProps.history, 'location.state.recordId'),
    };
    pushHistoryWithPrePage(ownProps.history, urlPath, state);
  };

  const onRowClick = async (expenseType: ExpenseType) => {
    const { history } = ownProps;
    const target = _.get(history, 'location.state.target');

    const isRowGroup = expenseType.isGroup;
    const isJorudanRecord =
      expenseType.recordType === RECORD_TYPE.TransitJorudanJP;
    const isMileageRec = isMileageRecord(expenseType.recordType);
    const recordId = _.get(history, 'location.state.recordId');

    const initValues = {
      recordDate: accountingDate,
      reportId,
      reportTypeId,
      items: [
        {
          expTypeId: expenseType.id,
          remarks: '',
          currencyId: '',
          taxTypeBaseId: '',
          taxTypeHistoryId: '',
          taxTypeName: '',
        },
      ],
    };
    // Records created through Add Record button in record list and report detail screen
    switch (true) {
      case isRowGroup:
        if (ownProps.recordType === RECORD_TYPE_CATEGORY.all) {
          // @ts-ignore
          dispatch(formValueAction.save(initValues));
        }

        pushHistoryWithPrePage(
          history,
          `${ownProps.history.location.pathname}/${expenseType.id}`,
          { target, recordId }
        );
        break;
      case isJorudanRecord:
        const routeForm = _.cloneDeep(formValues);

        routeForm.expenseType = expenseType.name;
        routeForm.expenseTypeId = expenseType.id;
        dispatch(routeFormPageActions.save(routeForm));

        if (recordId) {
          // existing report record
          history.push(`/request/report/route/edit/${recordId}/${reportId}`);
        } else {
          // new report record
          dispatch(
            // @ts-ignore
            formValueAction.save({
              ..._.cloneDeep(newRecord('', '', expenseType.recordType)),
              ...initValues,
            })
          );
          history.push(`/request/report/route/new`, { target });
        }

        break;
      default: // generate new record if change expense type
        // Initialize record with report header date or today when added through report detail or record list page
        const toForeignCurrency = expenseType.useForeignCurrency;
        const finalFormValues = {
          ...initValues,
          ..._.cloneDeep(formValues),
          ...(isMileageRec
            ? { mileageRouteInfo: { destinations: initialDestinations } }
            : {}),
        };

        if (isMileageRec) dispatch(mileageActions.setIsGeneratedPreview(false));

        const isExpTypeChanged =
          finalFormValues.items[0].expTypeId !== expenseType.id;
        let record = finalFormValues;

        if (isExpTypeChanged) {
          record = newRecord(
            expenseType.id,
            expenseType.name,
            expenseType.recordType,
            expenseType.useForeignCurrency,
            expenseType,
            true,
            expenseType.fileAttachment,
            expenseType.fixedForeignCurrencyId,
            expenseType.foreignCurrencyUsage,
            0,
            finalFormValues.recordDate,
            null,
            null,
            expenseType.merchant
          );
          record.items[0].remarks = finalFormValues.items[0].remarks;
          record.items[0].allowNegativeAmount = expenseType.allowNegativeAmount;

          if (finalFormValues.reportId) {
            record.reportId = finalFormValues.reportId;
          }

          if (toForeignCurrency) {
            // keep selected currency
            record.items[0].currencyId =
              expenseType.fixedForeignCurrencyId ||
              finalFormValues.items[0].currencyId;
          }
          record.recordId = finalFormValues.recordId;
          record.items[0].itemId = finalFormValues.items[0].itemId; // Temporary fix, change RecordItem for multiple items

          // set jct invoice option and clear jct registration number
          if (
            expenseType.jctRegistrationNumberUsage &&
            useJctRegistrationNumber
          ) {
            record.jctRegistrationNumberUsage =
              expenseType.jctRegistrationNumberUsage;
            if (isUseJctNo(expenseType.jctRegistrationNumberUsage)) {
              record.items[0].jctInvoiceOption = JCT_NUMBER_INVOICE.Invoice;
            }
          }

          // copy all previous extended item values on change of expense type
          Object.keys(finalFormValues.items[0]).forEach((x) => {
            if (
              x.includes('extendedItem') ||
              x.includes('costCenter') ||
              x.includes('job')
            ) {
              record.items[0][x] = finalFormValues.items[0][x];
            }
          });

          if (expenseType.fileAttachment !== RECEIPT_TYPE.NotUsed) {
            const { receiptList } = finalFormValues;
            record.receiptList = receiptList;
          }
        }

        if (!isExpTypeChanged) {
          // ocr detail manually input by user
          const amount = 0;

          const targetDate = finalFormValues.recordDate;

          record = newRecord(
            expenseType.id,
            expenseType.name,
            expenseType.recordType,
            expenseType.useForeignCurrency,
            expenseType,
            true,
            expenseType.fileAttachment,
            expenseType.fixedForeignCurrencyId,
            expenseType.foreignCurrencyUsage,
            amount,
            targetDate,
            undefined,
            undefined,
            expenseType.merchant
          );
          record = {
            ...record,
          };
          record.items[0] = {
            ...record.items[0],
            allowNegativeAmount: expenseType.allowNegativeAmount,
          };

          if (finalFormValues.reportId) {
            record.reportId = finalFormValues.reportId;
          }
          if (toForeignCurrency) {
            record.items[0].currencyId = expenseType.fixedForeignCurrencyId;
          }
        }

        if (
          expenseType.jctRegistrationNumberUsage &&
          useJctRegistrationNumber
        ) {
          record.jctRegistrationNumberUsage =
            expenseType.jctRegistrationNumberUsage;
          if (isUseJctNo(expenseType.jctRegistrationNumberUsage)) {
            record.items[0].jctInvoiceOption = JCT_NUMBER_INVOICE.Invoice;
          }
        }

        // get & set payment method option list for new record
        const paymentMethodOptionList = await dispatch(
          getPaymentMethodOptionList(record, isJorudanRecord)
        );
        record.paymentMethodId =
          _.get(paymentMethodOptionList, '0.value') || null;
        dispatch(formValueAction.save(record));
        // redirect
        if (record.recordId) {
          const url = formValues.reportId
            ? `/request/record/detail/${formValues.reportId}/${record.recordId}`
            : `/request/record/detail/${record.recordId}`;
          pushHistoryWithPrePage(ownProps.history, url, {
            isExpTypeChanged,
          });
        } else {
          const targetReport = 'report';
          const path = `/request/report/record/new/general`;
          pushHistoryWithPrePage(ownProps.history, path, {
            isExpTypeChanged,
            target: targetReport,
          });
        }

        break;
    }
  };

  const onBackClick = () => {
    goBack(ownProps.history);
  };

  return (
    <ExpTypeComponent
      keyword={ownProps.keyword}
      hasMore={hasMore}
      expenseTypeList={expenseTypeList}
      selectedCCTransaction={undefined}
      selectedOCRReceipt={undefined}
      onBackClick={onBackClick}
      onSearchClick={onSearchClick}
      onRowClick={onRowClick}
    />
  );
};

export default ExpenseTypeContainer;
