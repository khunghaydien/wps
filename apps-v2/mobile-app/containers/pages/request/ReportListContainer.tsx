import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { RouteComponentProps } from 'react-router-dom';
import { AnyAction } from 'redux';
import { ThunkDispatch } from 'redux-thunk';

import isEmpty from 'lodash/isEmpty';
import get from 'lodash/get';

import AppPermissionUtil from '../../../../commons/utils/AppPermissionUtil';
import { showToast } from '@commons/modules/toast';

import {
  REPORT_PER_PAGE_MOBILE,
  ReportIdList,
} from '../../../../domain/models/exp/Report';

import { State } from '../../../modules';
import { actions as ReportActions } from '../../../modules/expense/entities/report';
import { actions as reportIdListActions } from '../../../modules/expense/entities/reportIdList';
import { actions as reportListActions } from '../../../modules/expense/entities/reportList';
import {
  actions as listTypeUIActions,
  LIST_TYPE,
} from '@apps/mobile-app/modules/expense/ui/report/listType';

import { getReportIdList } from '../../../action-dispatchers/expense/ReportIdList';
import { getReportList } from '../../../action-dispatchers/expense/ReportList';

import ReportListPage from '../../../components/pages/expense/Report/List';

type OwnProps = RouteComponentProps;

const ReportListContainer = (ownProps: OwnProps) => {
  const dispatch = useDispatch() as ThunkDispatch<unknown, void, AnyAction>;

  const reportListType = useSelector(
    (state: State) => state.expense.ui.report.listType
  );
  const reportList = useSelector(
    (state: State) => state.expense.entities.reportList
  );
  const reportIdList = useSelector(
    (state: State) => state.expense.entities.reportIdList
  );
  const userSetting = useSelector((state: State) => state.userSetting);
  const employeeDetails = useSelector(
    (state: State) => state.common.exp.entities.employeeDetails
  );

  const { useExpense, employeeId, currencyId } = userSetting;
  const empHistories = get(employeeDetails, 'details');

  const hasPermissionError = AppPermissionUtil.checkPermissionError(
    useExpense,
    employeeId,
    currencyId
  );

  useEffect(() => {
    if (!isEmpty(empHistories)) {
      if (reportListType === LIST_TYPE.Approved) {
        onClickApprovedList();
      } else {
        boundGetReportList(null, false, REPORT_PER_PAGE_MOBILE, true)
          .then(() => {
            dispatch(
              getReportIdList(
                false,
                employeeId,
                undefined,
                true,
                getCompanyHistories()
              )
            );
          })
          .catch((_err) => false);
      }
    }

    return () => {
      dispatch(reportIdListActions.clear());
      dispatch(reportListActions.clear());
    };
  }, [empHistories]);

  const getCompanyHistories = () => {
    if (!isEmpty(empHistories))
      return empHistories.filter((h) => h.primary).map((h) => h.id);
    return [];
  };

  const onClickNewReport = () => {
    dispatch(ReportActions.clear());
    ownProps.history.push(`/request/report/new`);
  };

  const openDetail = (reportId: string, status: string, requestId?: string) => {
    ownProps.history.push(`/request/report/detail/${reportId}/${requestId}`, {
      status,
    });
  };

  const onClickApprovedList = () => {
    dispatch(listTypeUIActions.set(LIST_TYPE.Approved));
    dispatch(
      getReportIdList(true, employeeId, false, true, getCompanyHistories())
    )
      // @ts-ignore
      .then(([ids]) => {
        if (!isEmpty(ids)) {
          const idsToPass = ids.slice(0, REPORT_PER_PAGE_MOBILE);
          return (
            boundGetReportList(idsToPass, false, null, true)
              // @ts-ignore
              .catch(() => {
                dispatch(reportIdListActions.clear());
                dispatch(reportListActions.clear());
              })
          );
        } else {
          dispatch(reportListActions.clear());
        }
      });
  };

  const onClickBack = () => {
    dispatch(listTypeUIActions.set(LIST_TYPE.Active));
    boundGetReportList(null, false, REPORT_PER_PAGE_MOBILE, true)
      .then(() => {
        dispatch(
          getReportIdList(
            false,
            employeeId,
            undefined,
            true,
            getCompanyHistories()
          )
        );
      })
      .catch((err) => {
        const errMsg =
          (err.message && ` (${err.message})`) ||
          (err.event && ` (${err.event.message})`) ||
          '';
        dispatch(showToast(errMsg));
      });
  };

  const onRefresh = () => {
    const isApprovedList = reportListType === LIST_TYPE.Approved;
    return (
      dispatch(
        getReportIdList(
          isApprovedList,
          employeeId,
          true,
          true,
          getCompanyHistories()
        )
      )
        // @ts-ignore
        .then((ids: ReportIdList) => {
          if (!isEmpty(ids)) {
            return (
              boundGetReportList(
                ids.slice(0, REPORT_PER_PAGE_MOBILE),
                true,
                null,
                true
              )
                // @ts-ignore
                .catch(() => {
                  dispatch(reportIdListActions.clear());
                  dispatch(reportListActions.clear());
                })
            );
          } else {
            dispatch(reportListActions.clear());
          }
        })
    );
  };

  const boundGetReportList = (
    reportIds: ReportIdList,
    loadInBackground: boolean,
    noOfReport?: number,
    isRefresh?: boolean
  ) =>
    dispatch(
      getReportList(
        employeeId,
        reportIds,
        loadInBackground,
        noOfReport,
        isRefresh,
        true,
        getCompanyHistories()
      )
    )
      // @ts-ignore
      .then((res: Promise<ReportList>) => res);

  return (
    <ReportListPage
      userSetting={userSetting}
      reportList={reportList}
      reportIdList={reportIdList}
      hasPermissionError={hasPermissionError}
      reportListType={reportListType}
      openDetail={openDetail}
      onClickNewReport={onClickNewReport}
      onRefresh={onRefresh}
      onClickApprovedList={onClickApprovedList}
      onClickBack={onClickBack}
      getReportList={boundGetReportList}
      isRequest
    />
  );
};

export default ReportListContainer;
