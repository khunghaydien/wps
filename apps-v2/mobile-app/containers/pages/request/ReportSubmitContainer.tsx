import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { RouteComponentProps } from 'react-router-dom';
import { AnyAction } from 'redux';
import { ThunkDispatch } from 'redux-thunk';

import msg from '@commons/languages';
import Component from '@mobile/components/pages/commons/SubmitWithCommentPage';

import { State } from '@mobile/modules';
import { actions as ReportActions } from '@mobile/modules/expense/entities/report';

import { submit } from '@mobile/action-dispatchers/expense/ReportDetail';

type OwnProps = RouteComponentProps;

const ReportSubmitContainer = (ownProps: OwnProps) => {
  const dispatch = useDispatch() as ThunkDispatch<unknown, void, AnyAction>;

  const avatarUrl = useSelector((state: State) => state.userSetting.photoUrl);
  const reportId = useSelector(
    (state: State) => state.expense.entities.report.reportId
  );
  const empId = useSelector((state: State) => state.userSetting.employeeId);

  const title = msg().Appr_Lbl_SubmitComment;
  const submitLabel = msg().Appr_Lbl_Submit;

  const onClickBack = () => {
    ownProps.history.replace(`/request/report/detail/${reportId}`);
  };

  const onClickSubmit = (comment: string) => {
    dispatch(submit(reportId, comment, empId, true))
      // @ts-ignore
      .then(() => {
        ownProps.history.replace('/request/report/list');
        dispatch(ReportActions.clear());
      })
      .catch(() => {
        ownProps.history.replace(`/request/report/detail/${reportId}`);
      });
  };

  const getBackLabel = () => {
    return msg().Exp_Lbl_Report;
  };

  return (
    <Component
      avatarUrl={avatarUrl}
      title={title}
      submitLabel={submitLabel}
      onClickBack={onClickBack}
      onClickSubmit={onClickSubmit}
      getBackLabel={getBackLabel}
    />
  );
};

export default ReportSubmitContainer;
