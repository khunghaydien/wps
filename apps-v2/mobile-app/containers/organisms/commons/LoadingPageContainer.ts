import { connect } from 'react-redux';

import { $PropertyType } from 'utility-types';

import Component from '../../../components/organisms/commons/LoadingPage';
import { isShowing } from '../../../modules/commons/loading';

import { State } from '../../../modules';

type Props = $PropertyType<Component, 'props'>;

const mapStateToProps = (state: State, ownProps): Props => ({
  ...ownProps,
  isShowing: isShowing(state.mobileCommons.loading),
});

export default connect(mapStateToProps)(
  Component as React.ComponentType<Props>
);
