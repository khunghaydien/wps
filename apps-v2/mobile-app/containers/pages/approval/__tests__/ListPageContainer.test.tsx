import * as React from 'react';
import { Provider } from 'react-redux';
import { RouteComponentProps } from 'react-router-dom';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import { cloneDeep } from 'lodash';

import { act, render } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';

import { showAlert } from '@mobile/modules/commons/alert';
import { showConfirm } from '@mobile/modules/commons/confirm';

import RequestListRepository from '@apps/repositories/approval/RequestListRepository';
import RequestRepository from '@apps/repositories/approval/RequestRepository';

import { REQUEST_TYPE } from '@apps/domain/models/approval/request/Request';

import Component from '@mobile/components/pages/approval/ListPage';

import Container, {
  // @ts-ignore
  __get__,
  // @ts-ignore
  __set__,
  CHECKED_MAX,
} from '../ListPageContainer';

jest.mock('uuid/v4', () => () => 'uuid/v4/mock');
jest.mock('../../../../components/pages/approval/ListPage', () =>
  jest.fn(() => <div></div>)
);
jest.mock('../../../../modules/commons/confirm', () => ({
  __esModule: true,
  showConfirm: jest.fn(),
}));
jest.mock('../../../../modules/commons/alert', () => ({
  __esModule: true,
  showAlert: jest.fn(),
}));
jest.mock('../../../../../repositories/approval/RequestListRepository');
jest.mock('../../../../../repositories/approval/RequestRepository');

const router = {
  history: {
    push: jest.fn(),
  },
} as unknown as RouteComponentProps;

const MockComponent = Component as unknown as jest.Mock;

const dummyState = {
  approval: {
    entities: {
      list: {
        filterType: REQUEST_TYPE.ATTENDANCE_FIX,
        filteredList: [],
        targetList: [],
        checked: [],
      },
    },
    ui: {
      list: {
        comment: 'comment',
      },
    },
  },
  common: {
    accessControl: {
      permission: {
        canBulkApproveAttDailyRequest: true,
        canBulkApproveAttRequest: true,
      },
    },
  },
};

const mockStore = configureMockStore([thunk]);

const renderComponent = ({ back }: { back: boolean }, store) => {
  return render(
    <Provider store={store}>
      <Container {...router} back={back} />
    </Provider>
  );
};

beforeEach(() => {
  jest.clearAllMocks();
});
describe('props', () => {
  const mockUseCases = {
    load: jest.fn(),
    reload: jest.fn(),
    approve: jest.fn(),
  };

  const reverts = [];
  beforeAll(() => {
    reverts.push(__set__('useUseCases', () => mockUseCases));
    reverts.push(__set__('useFilterTypeOptions', () => []));
  });
  afterAll(() => {
    reverts.forEach((revert) => {
      revert();
    });
  });
  test('onClickRefresh.', () => {
    // Arrange
    const store = mockStore(dummyState);
    const _ = renderComponent({ back: false }, store);
    const props = MockComponent.mock.calls[0][0];
    store.clearActions();

    // Act
    props.onClickRefresh();

    // Assert
    expect(router.history.push).toBeCalledWith('/approval/list');
  });
  test('onChangeFilter.', () => {
    // Arrange
    const store = mockStore(dummyState);
    const _ = renderComponent({ back: false }, store);
    const props = MockComponent.mock.calls[0][0];
    store.clearActions();

    // Act
    props.onChangeFilter(REQUEST_TYPE.ATTENDANCE_FIX);

    // Assert
    expect(store.getActions()).toMatchSnapshot();
  });
  test('onCheckAll.', () => {
    // Arrange
    const store = mockStore(dummyState);
    const _ = renderComponent({ back: false }, store);
    const props = MockComponent.mock.calls[0][0];
    store.clearActions();

    // Act
    props.onCheckAll();

    // Assert
    expect(store.getActions()).toMatchSnapshot();
  });
  describe('onCheck', () => {
    test('when has record.', () => {
      // Arrange
      const state = cloneDeep(dummyState);
      state.approval.entities.list.targetList = [
        {
          recordId: 'xxxx',
          requestType: REQUEST_TYPE.ATTENDANCE_DAILY,
        },
      ];
      const store = mockStore(state);
      const _ = renderComponent({ back: false }, store);
      const props = MockComponent.mock.calls[0][0];
      store.clearActions();

      // Act
      props.onCheck({
        id: 'xxxx',
        requestType: REQUEST_TYPE.ATTENDANCE_DAILY,
      });

      // Assert
      expect(store.getActions()).toMatchSnapshot();
    });
    test("when hasn't record.", () => {
      // Arrange
      const state = cloneDeep(dummyState);
      state.approval.entities.list.targetList = [];
      const store = mockStore(state);
      const _ = renderComponent({ back: false }, store);
      const props = MockComponent.mock.calls[0][0];
      store.clearActions();

      // Act
      props.onCheck({
        requestId: 'xxxx',
        requestType: REQUEST_TYPE.ATTENDANCE_DAILY,
      });

      // Assert
      expect(store.getActions()).toMatchSnapshot();
      expect(router.history.push).toBeCalledWith(
        `/approval/list/select/${REQUEST_TYPE.ATTENDANCE_DAILY}/xxxx`
      );
    });
    test('when over max.', () => {
      // Arrange
      (showAlert as unknown as jest.Mock).mockReturnValueOnce({
        type: 'showAlert',
      });
      const state = cloneDeep(dummyState);
      state.approval.entities.list.targetList = [
        ...Array(CHECKED_MAX + 10).keys(),
      ].map((idx) => ({
        requestId: `${idx + 1}`,
        requestType: REQUEST_TYPE.ATTENDANCE_DAILY,
      }));
      state.approval.entities.list.checked = [...Array(CHECKED_MAX).keys()].map(
        (idx) => `${idx + 1}`
      );
      const store = mockStore(state);
      const _ = renderComponent({ back: false }, store);
      const props = MockComponent.mock.calls[0][0];
      store.clearActions();

      // Act
      props.onCheck({
        requestId: '101',
        requestType: REQUEST_TYPE.ATTENDANCE_DAILY,
      });

      // Assert
      expect(store.getActions()).toMatchSnapshot();
    });
  });
  test('onChangeComment.', () => {
    // Arrange
    const store = mockStore(dummyState);
    const _ = renderComponent({ back: false }, store);
    const props = MockComponent.mock.calls[0][0];
    store.clearActions();

    // Act
    props.onChangeComment('new comment');

    // Assert
    expect(store.getActions()).toMatchSnapshot();
  });
  describe('onClickApproveButton', () => {
    test('answer is true.', async () => {
      // Arrange
      const state = cloneDeep(dummyState);
      state.approval.entities.list.checked = ['1', '2', '3'];
      const store = mockStore(state);
      await act(async () => {
        renderComponent({ back: false }, store);
      });
      const props = MockComponent.mock.calls[0][0];
      mockUseCases.approve.mockResolvedValueOnce(true);
      store.clearActions();
      mockUseCases.load.mockClear();

      // Act
      await props.onClickApproveButton();

      // Assert
      expect(mockUseCases.approve).toBeCalledWith({
        ids: ['1', '2', '3'],
        comment: dummyState.approval.ui.list.comment,
      });
      expect(mockUseCases.load).toBeCalledTimes(1);
      expect(store.getActions()).toMatchSnapshot();
    });
    test('answer is false.', async () => {
      // Arrange
      const state = cloneDeep(dummyState);
      state.approval.entities.list.checked = ['1', '2', '3'];
      const store = mockStore(state);
      await act(async () => {
        renderComponent({ back: false }, store);
      });
      const props = MockComponent.mock.calls[0][0];
      mockUseCases.approve.mockResolvedValueOnce(false);
      store.clearActions();
      mockUseCases.load.mockClear();

      // Act
      await props.onClickApproveButton();

      // Assert
      expect(mockUseCases.approve).toBeCalledWith({
        ids: ['1', '2', '3'],
        comment: dummyState.approval.ui.list.comment,
      });
      expect(mockUseCases.load).toBeCalledTimes(0);
      expect(store.getActions()).toMatchSnapshot();
    });
  });
});

describe('useUseCases', () => {
  const store = mockStore(dummyState);
  const { result } = renderHook(() => __get__('useUseCases')(), {
    wrapper: ({ children }) => <Provider store={store}>{children}</Provider>,
  });
  const useCases = result.current;
  beforeEach(() => {
    store.clearActions();
  });
  test('load', async () => {
    // Arrange
    (RequestListRepository.fetch as unknown as jest.Mock).mockResolvedValueOnce(
      'approval list'
    );

    // Act
    await useCases.load();

    // Arrange
    expect(store.getActions()).toMatchSnapshot();
  });
  describe('approve', () => {
    test('answer is yes', async () => {
      // Arrange
      (RequestRepository.approve as unknown as jest.Mock).mockResolvedValueOnce(
        'approval list'
      );
      (showConfirm as unknown as jest.Mock).mockReturnValue(async () => true);

      // Act
      const result = await useCases.approve({
        ids: ['1', '2', '3'],
        comment: 'comment',
      });

      // Arrange
      expect(result).toBe(true);
      expect(RequestRepository.approve).toBeCalledTimes(1);
      expect(RequestRepository.approve).toBeCalledWith({
        ids: ['1', '2', '3'],
        comment: 'comment',
      });
      expect(store.getActions()).toMatchSnapshot();
    });
    test('answer is no', async () => {
      // Arrange
      (RequestRepository.approve as unknown as jest.Mock).mockResolvedValueOnce(
        'approval list'
      );
      (showConfirm as unknown as jest.Mock).mockReturnValue(async () => false);

      // Act
      const result = await useCases.approve({
        ids: ['1', '2', '3'],
        comment: 'comment',
      });

      // Arrange
      expect(result).toBe(false);
      expect(RequestRepository.approve).toBeCalledTimes(0);
      expect(store.getActions()).toMatchSnapshot();
    });
  });
});
