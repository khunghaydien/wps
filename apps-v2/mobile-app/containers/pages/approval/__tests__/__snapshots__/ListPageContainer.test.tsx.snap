// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`props onChangeComment. 1`] = `
Array [
  Object {
    "payload": "new comment",
    "type": "MOBILE/MODULES/APPROVAL/UI/LIST/SET_COMMENT",
  },
]
`;

exports[`props onChangeFilter. 1`] = `
Array [
  Object {
    "payload": "attendance_fix",
    "type": "MODULES/ENTITIES/APPROVAL/SET_FILTER",
  },
  Object {
    "type": "MODULES/ENTITIES/APPROVAL/CHECK_ALL",
  },
]
`;

exports[`props onCheck when has record. 1`] = `
Array [
  Object {
    "payload": Array [
      undefined,
    ],
    "type": "MODULES/ENTITIES/APPROVAL/TOGGLE",
  },
]
`;

exports[`props onCheck when hasn't record. 1`] = `Array []`;

exports[`props onCheck when over max. 1`] = `
Array [
  Object {
    "type": "showAlert",
  },
]
`;

exports[`props onCheckAll. 1`] = `
Array [
  Object {
    "type": "MODULES/ENTITIES/APPROVAL/TOGGLE_ALL",
  },
]
`;

exports[`props onClickApproveButton answer is false. 1`] = `Array []`;

exports[`props onClickApproveButton answer is true. 1`] = `Array []`;

exports[`useUseCases approve answer is no 1`] = `Array []`;

exports[`useUseCases approve answer is yes 1`] = `
Array [
  Object {
    "payload": "uuid/v4/mock",
    "type": "START_LOADING",
  },
  Object {
    "type": "MOBILE/MODULES/APPROVAL/UI/LIST/INITIALIZE",
  },
  Object {
    "payload": Object {
      "message": "Approved",
      "options": undefined,
      "variant": "success",
    },
    "type": "COMMONS/MODULES/TOAST/SHOW",
  },
  Object {
    "payload": "uuid/v4/mock",
    "type": "END_LOADING",
  },
]
`;

exports[`useUseCases load 1`] = `
Array [
  Object {
    "payload": "uuid/v4/mock",
    "type": "START_LOADING",
  },
  Object {
    "payload": "approval list",
    "type": "MODULES/ENTITIES/APPROVAL/SET_RECORDS",
  },
  Object {
    "payload": "uuid/v4/mock",
    "type": "END_LOADING",
  },
]
`;
